engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/blueprint/IntrospectionService.kt

```diff
@@ -0,0 +1,29 @@
+package graphql.nadel.enginekt.blueprint
+
+import graphql.ExecutionInput
+import graphql.GraphQL
+import graphql.GraphqlErrorHelper.toSpecification
+import graphql.language.AstPrinter
+import graphql.nadel.Service
+import graphql.nadel.ServiceExecution
+import graphql.nadel.ServiceExecutionResult
+import graphql.schema.GraphQLSchema
+
+internal class IntrospectionService constructor(schema: GraphQLSchema) :
+    Service(name, schema, makeServiceExecution(schema), null, null) {
+
+    companion object {
+        const val name = "__introspection"
+
+        private fun makeServiceExecution(schema: GraphQLSchema): ServiceExecution {
```

Comment:
I think we need to provide a default impl and feed it in as an argument to `Nadel` or `NextgenEngine`.

We need to do this because the AGG needs to wrap around this code to time the introspection, for metrics sake and for internal latency.

e.g.

```kotlin
internal class IntrospctionService(
  schema: GraphQLSchema,
  introspectionRunnerFactory: IntrospectionRunnerFactory,
) : Service(name, schema, introspectionRunnerFactory.make(schema), null, null)

fun interface IntrospectionRunnerFactory {
  fun make(schema: GraphQLSchema): ServiceExecution
}

open class DefaultIntrospectionRunner(schema: GraphQLSchema): ServiceExecution {
  private val graphQL = GraphQL.newSchema(schema).build()

  override fun execute(params): CompletableFuture<ServiceExecutionResult> {
    return graphQL.executeAsync(…).thenApply { … }
  }
}
```

Then the AGG can be like

```kotlin
class GatewayIntrospectionRunner(schema: GraphQLSchema) : DefaultIntrospectionRunner(schema) {
  override fun execute(params): CompletableFuture {
    // start timer
    return super.execute(params).whenComplete((result, throwable) -> /* record internal latency etc. */)
  }
}
```

Reply:
interesting. does the current engine provide this functionality or is it a new feature you think about introducing?

Reply:
Current engine supports this. We use it to record a time to execute introspection metric.

See `GatewayIntrospectionRunner`. We should also be using the new class to exclude introspection times from internal latency. The way we do it in the AGG won't work going forward. We currently simply don't record internal latency if there's zero external calls. This works because currently we don't allow mixing of introspection and normal fields. So we NEED this functionality for nextgen.